presto-main/src/main/java/com/facebook/presto/metadata/FunctionListBuilder.java

```diff
@@ -149,9 +159,65 @@ public FunctionListBuilder scalar(Class<?> clazz)
         catch (IllegalAccessException e) {
             throw Throwables.propagate(e);
         }
+        groupMatchingScalars();
+        return this;
+    }
+
+    private FunctionListBuilder groupMatchingScalars()
+    {
+        List<ParametricFunction> newFunctions = groupMatchingScalars(functions);
+        functions.clear();
+        functions.addAll(newFunctions);
         return this;
     }
 
+    private List<ParametricFunction> groupMatchingScalars(List<ParametricFunction> inputFunctions)
+    {
+        inputFunctions.forEach(f -> checkArgument(f instanceof FunctionInfo, "expected FunctionInfo but got %s", f));
+        Map<Signature, List<FunctionInfo>> groupedFunctions = inputFunctions.stream()
+                .map(f -> (FunctionInfo) f)
+                .collect(Collectors.groupingBy(ParametricFunction::getSignature));
+        List<ParametricFunction> resultFunctions = new ArrayList<>();
+        for (Map.Entry<Signature, List<FunctionInfo>> entry : groupedFunctions.entrySet()) {
+            List<FunctionInfo> functionsGroup = entry.getValue();
+            if (functionsGroup.size() == 1) {
+                resultFunctions.add(getOnlyElement(functionsGroup));
+            }
+            else {
+                resultFunctions.add(buildGroupingScalarWrapper(functionsGroup));
+            }
+        }
+        return resultFunctions;
+    }
+
+    private ParametricFunction buildGroupingScalarWrapper(List<FunctionInfo> functionsGroup)
+    {
+        checkArgument(functionsGroup.size() > 1, "functions group must have multiple elements");
```

Comment:
The code above makes sure this is not called for single element lists.
It would work though.